model-optimizer/extensions/ops/sparse_reshape.py

```diff
@@ -28,26 +30,41 @@ def supported_attrs(self):
 
     @staticmethod
     def infer(node: Node):
+        name = node.soft_get('name', node.id)
+
         input_indices_shape = node.in_port(0).data.get_shape()
         input_indices_value = node.in_port(0).data.get_value()
-        input_shape_value = node.in_port(1).data.get_value()
-        new_shape_value = node.in_port(2).data.get_value()
+        input_shape = node.in_port(1).data.get_value()
+        new_shape = node.in_port(2).data.get_value()
         new_shape_shape = node.in_port(2).data.get_shape()
 
-        assert input_shape_value is not None and new_shape_value is not None, \
+        assert input_shape is not None and new_shape is not None, \
             "Values for input shape and new shape must be defined"
-        assert np.count_nonzero(new_shape_value == -1) <= 1, \
+        assert len(np.argwhere(new_shape == -1)) <= 1, \
             "Value -1 occurs in new shape value more than once"
+        assert len(np.argwhere(new_shape < -1)) == 0, \
+            "Only non-negative or -1 values are allowed"
+
+        output_shape = np.ma.masked_array(new_shape, mask=new_shape == -1)
+        assert not is_fully_defined(input_shape) or not is_fully_defined(output_shape) or \
+               np.prod(input_shape) == np.prod(output_shape), \
+            "Number of elements in input {} and output {} of dynamic reshape node {} mismatch" \
+            "".format(input_shape, output_shape, name)
+
+        # we can deduce -1 only if input_shape is fully defined and
+        # there is one dynamic dimension in output_shape
+        if is_fully_defined(input_shape) and np.ma.count_masked(output_shape) == 1:
+            undefined_dim_size = np.prod(input_shape) // np.prod(output_shape)
```

Comment:
`np.prod(output_shape)`: you need to compute prod of all elements except dynamic dimension

Reply:
please add test for this case

Reply:
@pavel-esir, please check this line:
` undefined_dim_size = np.prod(input_shape) // np.prod(output_shape)`
because `output_shape` constains exactly one dynamic_dimension_value but you need to exclude from the product